src/components/stopwatch/StopWatchController.cpp

```diff
@@ -0,0 +1,101 @@
+#include "components/stopwatch/StopWatchController.h"
+
+using namespace Pinetime::Controllers;
+
+namespace {
+  TickType_t CalculateDelta(const TickType_t startTime, const TickType_t currentTime) {
+    TickType_t delta = 0;
+    // Take care of overflow
+    if (startTime > currentTime) {
+      delta = 0xffffffff - startTime;
+      delta += (currentTime + 1);
+    } else {
+      delta = currentTime - startTime;
+    }
+    return delta;
+  }
+}
+
+StopWatchController::StopWatchController() {
+  Clear();
+}
+
+// State Change
+
+void StopWatchController::Start() {
+  currentState = StopWatchStates::Running;
+  startTime = xTaskGetTickCount();
+}
+
+void StopWatchController::Pause() {
+  currentState = StopWatchStates::Paused;
+  timeElapsedPreviously += CalculateDelta(startTime, xTaskGetTickCount());
+}
+
+void StopWatchController::Clear() {
+  currentState = StopWatchStates::Cleared;
+  timeElapsedPreviously = 0;
+
+  for (int i = 0; i < LAP_CAPACITY; i++) {
+    laps[i].count = 0;
+    laps[i].time = 0;
+  }
+  lapCount = 0;
+  lapHead = 0;
+}
+
+// Lap
+
+void StopWatchController::PushLap() {
+  TickType_t lapEnd = GetElapsedTime();
+  laps[lapHead].time = lapEnd;
+  laps[lapHead].count = lapCount + 1;
+  lapCount += 1;
+  lapHead = lapCount % LAP_CAPACITY;
+}
+
+int StopWatchController::GetLapNum() {
+  if (lapCount < LAP_CAPACITY)
+    return lapCount;
+  else
+    return LAP_CAPACITY;
+}
+
+int StopWatchController::GetLapCount() {
+  return lapCount;
+}
+
+int Wrap(int index) {
+  return ((index % LAP_CAPACITY) + LAP_CAPACITY) % LAP_CAPACITY;
+}
+
+LapInfo* StopWatchController::LastLap(int lap) {
+  if (lap >= LAP_CAPACITY || lap > lapCount || lapCount == 0) {
+    // Return "empty" LapInfo_t
+    return &emptyLapInfo;
+  }
+  // Index backwards
+  int index = Wrap(lapHead - lap);
+  return &laps[index];
+}
```

Comment:
Could [CircularBuffer](https://github.com/InfiniTimeOrg/InfiniTime/blob/main/src/utility/CircularBuffer.h) be used here? Maybe I'm misreading but it seems like this is implementing a similar thing

Reply:
It could certainly be used but I would leave that change for a PR that deals with lap times specifically. Lap time handling in StopWatch is quite basic right now, I suppose it will see more work in the future.

Reply:
Hmm OK, my thought process is that if we're implementing a brand new controller we may as well do it right rather than refactor it later? I haven't seen any push for change with lap handling, so I doubt this will see future work. If you don't fancy doing it yourself I wouldn't mind having a go at it, perhaps my intuition about it being suitable here is wrong anyway

Reply:
I looked at CircularBuffer, it wasn't that much work and it made the code much simpler. Thanks!